Projects: ad2s1210: Add support for ad2s1210

ahaslam2 opened the pull request:
The AD2S1210 is a complete 10-bit to 16-bit resolution tracking resolver-to-digital converter

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/precision-converters-firmware?pullRequest=24) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/precision-converters-firmware?pullRequest=24) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/precision-converters-firmware?pullRequest=24) it.</sub>

mphalke commented:
@ahaslam2 Please sign the CLA before we start reviewing this PR

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+	.extra = &uart_extra_init_params
+};
+
+
```
Extra line can be removed

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+{
+	return no_os_uart_init(&uart_desc, &uart_init_params);
+}
+/**
```
Blank line can be added above

ribdp commented on projects/ad2s1210_iio/app/main.c:
```diff
@@ -0,0 +1,51 @@
+#include <stdint.h>
+
+#include "ad2s1210_iio.h"
+#include "no_os_error.h"
```
nit: Are we using anything from this file? If not, it can be removed

ribdp commented on projects/ad2s1210_iio/app/app_config.h:
```diff
@@ -0,0 +1,154 @@
+#define AD2S1210_FCLKIN		8192000
+
+#define MATH_PI                   3.1415926f
+/* velocity scale depends on reolution */
```
nit: typo reolution -> resolution

ahaslam2 replied:
ok

ribdp commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+	uint8_t hysteresis;
+	int ret;
+
+	switch (priv) {
```
Is sampling_frequency a read-only attribute?

ahaslam2 replied:
ok

ahaslam2 replied:
yes, its currently hard coded to 16khz (as most projects i saw in this repo) i think this only controls the pwm frequency that triggers the irq where we sample.  so i suppose i could make it writeable too that will change the freq of the pwm. is it required? should i do this?

ribdp replied:
I think we can leave it be for now.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.h:
```diff
@@ -0,0 +1,37 @@
+
+#include "iio.h"
+#include "iio_types.h"
+#include "ad2s1210.h"
```
Are these includes needed? 
If they are required in the source file, should we just move them there?

ahaslam2 replied:
will remove.

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+
+/* UART descriptor */
+struct no_os_uart_desc *uart_desc;
+struct no_os_gpio_desc *trigger_gpio_desc;
```
Comments can be added for the other descriptors as well

ahaslam2 replied:
ok will do, but  not sure all these comments are very useful as the variable name is almost the same as the comment and self explanatory.

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+
+	/* Initialize the IRQ controller */
+	ret = no_os_irq_ctrl_init(&trigger_irq_desc, &trigger_gpio_irq_params);
+	if (ret)
```
Brackets can be added around if {}

ahaslam2 replied:
this is a single line return statement around the if? why add brackets? if its mandatory for coding style in this repo, should i add brackets to all the other single line after if statements?

jansunil replied:
Yes, the general coding style for this repository.

ahaslam2 replied:
ok.. i will have to add that everywhere..

jansunil commented on projects/ad2s1210_iio/app/app_config.h:
```diff
@@ -0,0 +1,154 @@
+#endif
+
+/* Select the RESOLVER data capture mode (default is CC mode) */
+#define DATA_CAPTURE_MODE CONTINUOUS_DATA_CAPTURE
```
Do we need this line?

ahaslam2 replied:
will remove.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+	.ch_type = _type,\
+	.ch_out = _ch_out,\
+	.indexed = true,\
+	.channel = 0,\
```
Should this be '_idx' ?

ahaslam2 replied:
Originally i had it as _idx, but if we want to be in par with the linux kernel driver attributes it should be 0. 

These are the linux kernel attributes, if you check the suffix of the channel is always 0. and i guess it makes some sense as its the position and velocity of one resolver device

			angl0:  (input, index: 0, format: be:U16/16>>0)
			5 channel-specific attributes found:
				attr  0: hysteresis value: 1
				attr  1: hysteresis_available value: 0 1
				attr  2: label value: position
				attr  3: raw value: 12579
				attr  4: scale value: 0.000095874
			anglvel0:  (input, index: 1, format: be:S16/16>>0)
			3 channel-specific attributes found:
				attr  0: label value: velocity
				attr  1: raw value: 9
				attr  2: scale value: 0.023968449
			timestamp:  (input, index: 2, format: le:S64/64>>0)
			phase0:  (input)
			1 channel-specific attributes found:
				attr  0: label value: synthetic reference
			altvoltage0:  (output)
			3 channel-specific attributes found:
				attr  0: frequency value: 10000
				attr  1: frequency_available value: [2000 250 20000]
				attr  2: label value: excitation

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		.realbits = CHN_STORAGE_BITS,
+		.storagebits = CHN_STORAGE_BITS,
+		.shift = 0,
+		.is_big_endian = false,
```
Do these channels have different polarity?

ahaslam2 replied:
yes. this is a output channel that represents the  excitation frequency we write on a register, and its all passed in little endian from the attributes.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		.realbits = CHN_STORAGE_BITS,
+		.storagebits = CHN_STORAGE_BITS,
+		.shift = 0,
+		.is_big_endian = false,
```
Any specific reason ``is_big_endian`` is different here?

ahaslam2 replied:
same comment as above. this is output channel, the value to set the frequency  to commes from the set attributes callback, and its all little endian.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+static bool buf_size_updated;
+
+static const float ad2s1210_velocity_scale[] = {
+	17089132, /* 8.192MHz / (2*pi * 2500 / 2^15) */
```
Can we add macros defined with the equation mentioned in the comment for these values?

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		return snprintf(buf, len, "position");
+
+	case HYSTERESIS_ATTR_ID:
+		ret = ad2s1210_hysteresis_is_enabled(dev);
```
Where is this function defined?

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		ret = ad2s1210_hysteresis_is_enabled(dev);
+		if (ret < 0)
+			return ret;
+		return snprintf(buf, len, "%d", ret);
```
Why are we printing the return status here? Curious to know what the ad2s1210_hysteresis_is_enabled() actually returns

ahaslam2 replied:
ad2s1210_hysteresis_is_enabled returns negative in case of error, 0 if hysteresis is disabled and 1 if its enabled.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+			else
+				chn = AD2S1210_VEL;
+
+			if (!(iio_dev_data->buffer->active_mask & (1 << i)))
```
Would it be a good idea to do these operations in the  iio_ad2s1210_prepare_transfer? That way, we need not repeat these operations for every cycle. It also reduces the time the processor would take to read one sample

ahaslam2 replied:
I dont see a super optimization here?. ad2s1210_spi_single_conversion will only read one "channel" at the time . (with this hw we cannot read the 2 channels with the same spi operation). is the bit check that you find expensive? AFAICT  we will always have to do one check anyways to see which channel is/are enabled to push the right data into the buffer,

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		uint32_t chn_mask)
+{
+#if (DATA_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
+
```
extra line can be removed

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+		buf_size_updated = true;
+	}
+
+	/* Push into circular buffer */
```
This comment can be moved to the line where  no_os_cb_write() is being called

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+	return 0;
+}
+
+#if (DATA_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
```
This check is not really needed.

ahaslam2 replied:
its to avoid compiler warinings in case BURST data capture is defined
[Warning] ad2s1210_iio.c@494,16: 'ad2s1210_iio_trigger_param_init' defined but not used [-Wunused-function]

jansunil replied:
Compiler warnings should not be an issue. This check is not really added in the other projects. hence the suggestion

jansunil commented on projects/ad2s1210_iio/app/app_config.h:
```diff
@@ -0,0 +1,154 @@
+/******************************************************************************/
+
+extern struct no_os_uart_desc *uart_desc;
+extern struct no_os_gpio_desc *led_gpio_desc;
```
Is this being used somewhere?

ahaslam2 replied:
no!.. will remove

jansunil commented on projects/ad2s1210_iio/app/app_config_mbed.h:
```diff
@@ -0,0 +1,59 @@
+#include <PinNames.h>
+
+#include "mbed_uart.h"
+#include "mbed_irq.h"
```
Do we need this include?

ahaslam2 replied:
no, will remove

jansunil commented on projects/ad2s1210_iio/app/app_config_mbed.h:
```diff
@@ -0,0 +1,59 @@
+#define UART_TX		CONSOLE_TX
+#define	UART_RX		CONSOLE_RX
+
+#define SAMPLING_RATE                                   (16000)
```
Is this the maximum sampling rate mentioned in the datasheet?
if not, a short comment can be added stating the reason why this is only 16ksps.

Example - https://github.com/analogdevicesinc/precision-converters-firmware/blob/main/projects/ad4696_iio/app/app_config_mbed.h#L50

ahaslam2 replied:
there is not max sampling rate mentioned in datasheet. i kind out find out a sane value max experimentally from my current setup with flywires + the breakout board. not sure exactly what the max value i  can use is.  
ill add a comment on how this value was experimentally  selected.

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+	.extra = &pwm_extra_init_params,
+	.platform_ops = &pwm_ops
+};
+/* UART init parameters */
```
A blank line can be added above

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+			if (!(iio_dev_data->buffer->active_mask & (1 << i)))
+				continue;
+
+			ret = ad2s1210_spi_single_conversion(ad2s1210_dev_inst,
```
How are we making sure that the sample is being read only after the End of conversion signal?
Continuous mode makes sure that the samples are read only after the falling edge in the PWM pin is detected. How is it done here for burst mode?

ahaslam2 replied:
the no-os driver will toggle the sample pin before a read.

jansunil replied:
So what is the difference between the GPIO_SAMPLE and the external PWM that we are triggering manually?
I am trying to understand how is the end of conversion being handled in continuous vs burst mode. 

Looks like ad2s1210_spi_single_conversion() is being used in  iio_ad2s1210_submit_buffer() and in the ad2s1210_trigger_handler(). So the functionality for reading the sample is the same. But how is end of conversion detected in burst mode?

ahaslam2 replied:
in the case of continues capture  pwm triggers the irq handler that will call ad2s1210_spi_single_conversion where the gpio is toggled before reading from the spi. IOW, the pwm does not directly toggle the sample pin. 

i the case of burst capture,there is no pwm, we call ad2s1210_spi_single_conversion in a loop the number of times that is needed to fill the buffer . each call, the sample gpio is toggled in the ad2s1210_spi_single_conversion function. when the gpio is toggled the sample is captured. after a small delay (350ns) of the toggle the sample can be read. right now i dont specifficaly add a delay in the no-os driver. but i should, and will do.

jansunil replied:
okay. Let us not add the extra delay

jansunil commented:
General question - EEPROM implementation is not seen. Are there plans to add it in the future?

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+	return 0;
+}
+#endif
+/**
```
A blank line can be added between two functions

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+
+#define AD2S1210_ENABLE_HYSTERESIS	NO_OS_BIT(4)
+
+enum ad2s1210_attribute_id {
```
Can we move this to the "variables/user defined data types" section?

ahaslam2 replied:
yes

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+        ((((x) & 0x00ff) << 8) | \
+         (((x) & 0xff00) >> 8))
+
+#define AD2S1210_ENABLE_HYSTERESIS	NO_OS_BIT(4)
```
Are we using this macro anywhere?

ahaslam2 replied:
removed. i defined this on the no-os driver .h patch instead.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+/* IIO interface descriptor */
+static struct iio_desc *ad2s1210_iio_desc;
+
+#if (DATA_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
```
We dont really need this check. The variable can be defined irrespective of what the data capture mode is

ahaslam2 replied:
but if i remove the check we get a complier warning about variable not used.
are we ok with such warnings in these projects?

jansunil replied:
I guess that should not be a problem. The other projects follow a similar structure, hence suggesting the same.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+ */
+static int32_t iio_ad2s1210_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+#if (DATA_CAPTURE_MODE == BURST_DATA_CAPTURE)
```
Macro check can be moved after declaration of variables

ahaslam2 replied:
ok with compiler  "not used" warnings?

jansunil commented on projects/ad2s1210_iio/app/app_config.c:
```diff
@@ -0,0 +1,166 @@
+
+/* UART descriptor */
+struct no_os_uart_desc *uart_desc;
+struct no_os_gpio_desc *trigger_gpio_desc;
```
Are we using this pin as a GPIO anywhere?

ahaslam2 replied:
yes. in gpio_trigger_Init.  i think the pwm output is used on the gpio pin (input) as an irq,
its all muxed internally in the soc. this triggers the sampling at pwm frequency.  is my understanding correct?

jansunil replied:
The understanding is correct, but any reason why we are specifically initializing the pin as a GPIO again?
PWM pin would be responsible for generating the signal at the frequency of set sampling rate, which would be initialized as a source of interrupt via the IRQ controller. Then do we really need the GPIO initialization for the same pin?

Please correct me if my understanding is wrong

ahaslam2 replied:
you are right. we dont need this gpio. Its the _gpio_ in gpio_irq that confused me a bit. i will remove this.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_user_config.c:
```diff
@@ -0,0 +1,65 @@
+		.platform_ops = &mbed_gpio_ops,
+		.extra = NULL
+	},
+	.resolution = AD2S1210_RES_16BIT,
```
Does this part allow the resolution to be configured to something other than 16?
If yes, it would be a good idea to add a macro in the app_config.h for selecting the same.
The macro can be used to assign the value of ``.resolution`` here as well

ahaslam2 replied:
yes we can use 16, 14,12,10. ill add the macro

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_user_config.c:
```diff
@@ -0,0 +1,65 @@
+/******************************************************************************/
+
+#include "ad2s1210_user_config.h"
+#include "app_config.h"
```
Do we need this include here?

ahaslam2 replied:
yes for AD2S1210_MAX_CLKIN, and RESOLUTION if we will add it in app_config.h (suggestion above)

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+#include <errno.h>
+
+#include "iio.h"
+#include "iio_types.h"
```
iio_types.h and iio.h are already included within the iio_trigger. 
Do we need it here?

ahaslam2 replied:
ok ill remove it. (as a side note: i usually would explicitly include the headers that are needed the c files to avoid depending on other h headers to include what is need. )

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+
+	case FREQ_AVAIL_ATTR_ID:
+		return snprintf(buf, len, "[%d %d %d]", AD2S1210_MIN_EXCIT,
+				250, AD2S1210_MAX_EXCIT);
```
What is the significance of the number -250?

ahaslam2 replied:
this is needed to be in par with the linux attributes (comes from linux driver). 250 is the step size for each selectable frequency between min and max. i will add a macro to make it clear.

jansunil commented:
General comment - .gitignore file needs to be added within the projects/ad2s1210_iio directory
Example- https://github.com/analogdevicesinc/precision-converters-firmware/blob/main/projects/ad4696_iio/.gitignore

ahaslam2 commented on projects/ad2s1210_iio/app/ad2s1210_user_config.c:
```diff
@@ -0,0 +1,65 @@
+		.extra = &spi_extra_init_params
+	},
+	.gpio_a0 = {
+		.number = ARDUINO_UNO_D0,
```
ok

ahaslam2 commented on projects/ad2s1210_iio/app/app_config.h:
```diff
@@ -0,0 +1,154 @@
+#define RESOLVER_MAX_ATTR		10
+
+/* Max count is always 16 bit. LSBs are ignored in lower resolutions */
+#define RESOLVER_MAX_COUNT_UNIPOLAR  (uint32_t)((1 << 16) - 1)
```
ok

ahaslam2 commented on projects/ad2s1210_iio/app/app_config_mbed.c:
```diff
@@ -0,0 +1,68 @@
+/******************************************************************************/
+
+#include <stdbool.h>
+#include "app_config.h"
```
this was carried form other projects too,  its needed for:
VIRTUAL_COM_PORT_VID, VIRTUAL_COM_SERIAL_NUM

But seems im missing .platform_ops = &vcom_ops, in the uart_init_params

any idea how i can see the console output from this VCOM port if enabled?

ahaslam2 commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+	.attributes = &ad2s1210_iio_ch_attributes[_idx][0]\
+}
+
+/*	Number of IIO devices */
```
ok

ahaslam2 commented:
Change log of last push:

Fix comments from @jansunil:
- various blank spaces and lines fixes
- removed iio_types.h included by iio_trigger.h
- removed AD2S1210_ENABLE_HYSTERESIS defined in no-os driver
- moved ad2s1210_attribute_id to user defined section
- uses macros for ad2s1210_velocity_scale table
- removed iio_types.h and ad2s1210.h from ad2s1210_iio.h
- used gpio defined in mbed_config for ad2s1210_user_config
- add resolution macro in app_config
- define and use vcom_ops as done in ad738x
- add descriptor comments
- use macro limits for max count
- remove unused led_gpio_desc
- add comment about how sample rate was defined

ahaslam2 commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,619 @@
+#if (DATA_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
+
+	int32_t ret;
+
```
As far as i can see, there is no such thing as  enable/disable. its a bit different than a conventional adc, its a resolver  with 2 values angle and angle velocity that are always on.

ahaslam2 commented:
Change log of last push:
- Added brackets for single line if statements.

ahaslam2 commented:
Changes on last  push

- use channel mask for single conversion
- less code in conversion loop

jansunil commented on projects/ad2s1210_iio/README.txt:
```diff
@@ -0,0 +1,43 @@
+Use of the Mbed platform allows code to be ported to other Mbed supported target boards with little or no modifications.
+
+Product details: AD2S1210.
+Product Evaluation board details: EVAL-AD2S1210SDZ
```
Do we have an arduino based board for this part? 
I see that the macros in the app_config_mbed.h maps the pins to arduino headers. So would the board ID here need an update?

ahaslam2 replied:
With current hardware available the setup is fly wires from the arduino pins on the sdp-k1 to a breakout board that is connected to the ad2s1210 eval. I need to describe this better on the readme. will do.

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,627 @@
+	AD2S1210_TRACKING_RATE_16BIT,
+};
+
+static int iio_ad2s1210_attr_get(void *device, char *buf, uint32_t len,
```
Comments for function header seem to be missing here

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,627 @@
+	return 0;
+}
+#else
+static int32_t iio_ad2s1210_submit_buffer(struct iio_device_data *iio_dev_data)
```
Why do we need two functions defined here? Compiler warnings should not be a problem for unused variables in such cases

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,627 @@
+	return 0;
+}
+
+int32_t ad2s1210_trigger_handler(struct iio_device_data *iio_dev_data)
```
Comments above function definition seem to be missing here

ahaslam2 replied:
ok

jansunil commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,627 @@
+
+	struct iio_device *iio_ad2s1210_dev;
+	/* IIO device descriptors */
+
```
Extra line can be removed

ahaslam2 commented:
Change log from last push
- document fly wire setup in readme. 
- added function doc for attr_set, attr_get and trigger_handler
- removed unneeded trigger_gpio_param
- reorder DATA_CAPTURE_MODE defines like other projects (ignore warnings)

@jansunil i think i addressed all the comments,   https://github.com/analogdevicesinc/no-OS/pull/1953
was merged.  if there are no more comments would it be ok to merge this PR now?

jansunil commented:
The changes look good to me. 
@mphalke @ribdp  - Do you have any comments on these commits?

jansunil commented:
@ahaslam2  - Looks like some APIs are dependent on the latest no-OS driver commits from github.com/analogdevicesinc/no-OS/pull/1953

Should we also update the commit hash here?

ribdp commented on projects/ad2s1210_iio/app/ad2s1210_iio.c:
```diff
@@ -0,0 +1,645 @@
+static int8_t data_buffer[DATA_BUFFER_SIZE] = { 0 };
+#endif
+
+#define swab16(x) \
```
Looks like a typo here (swap)

There's also this helper function in no-OS that can accomplish this by specifying a step size of 2

ahaslam2 replied:
not a typo, swab is the name used in  linux i had just picked the same name.
https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/swab.h
https://man7.org/linux/man-pages/man3/swab.3.html

that no os function looks more complex than just what this macro is doing, but ok
ill try it out and update the patch if it works

ribdp replied:
> not a typo, swab is the name used in linux i had just picked the same name.

Oh I see, wasn't aware of this...

ahaslam2 commented:
> @ahaslam2 - Looks like some APIs are dependent on the latest no-OS driver commits from github.com/[analogdevicesinc/no-OS/pull/1953](https://github.com/analogdevicesinc/no-OS/pull/1953)
> 
> Should we also update the commit hash here?

@jansunil  sorry not sure what i should update. you mean i should update the commit hash of libraries/no-OS with the latest commit hash of no-os?

ahaslam2 commented:
change log:

- use  no_os_memswap64 instead of swab macro. (@ribdp )

jansunil commented:
> > @ahaslam2 - Looks like some APIs are dependent on the latest no-OS driver commits from github.com/[analogdevicesinc/no-OS/pull/1953](https://github.com/analogdevicesinc/no-OS/pull/1953)
> > Should we also update the commit hash here?
> 
> @jansunil sorry not sure what i should update. you mean i should update the commit hash of libraries/no-OS with the latest commit hash of no-os?

@ahaslam2 So the latest commit that has been made to the ad2s1210 no-OS drivers is [here](https://github.com/analogdevicesinc/no-OS/commit/2627ec0bfe7f79ef3b0027b6a96ea2a4a7aa6d8c) so the no-OS.lib file and its respective sub-module needs to be referenced "2627ec0bfe7f79ef3b0027b6a96ea2a4a7aa6d8c" as its commit ID.

A straightforward copy-paste of this ID would be enough for updating the no-OS.lib
Please check [here](https://stackoverflow.com/questions/10914022/how-do-i-check-out-a-specific-version-of-a-submodule-using-git-submodule) for some inputs on how to update the sub-modules.

ahaslam2 commented:
Changelog:

- updated no-os submodule  commit hash to include dependent patch.

ahaslam2 commented:
@jansunil is it ok now ? Not sure what the CI build errors are about, they seem not related to this patch.
I tested 

- git clone --recursive https://github.com/analogdevicesinc/precision-converters-firmware
- git cherry-pick remotes/origin/ahaslam/ad2s1210_rc1 -1
- git submodule update --recursive --init

mbed-compile was OK.

ahaslam2 commented:
@jansunil @mphalke  i added a patch in this MR to update to the latest version of the libs:

https://github.com/analogdevicesinc/precision-converters-firmware/pull/24/commits/8d787723bfbfed2e1ee70b91da07602549b1da42

I compiled with mbed-cli all seems ok would it be possible you take another look at this and let me know if its ok to merge?

mphalke commented:
@ahaslam2 I had given some comments last month but unfortunately those were not get reflected into the PR. Apologies.
Please check them now and do the necessary updates to code.

mphalke commented on libraries/CMSIS-DSP.lib:
```diff
@@ -1 +1 @@
-https://github.com/ARM-software/CMSIS-DSP/#3a04f817a4380c00ff51f9bcdc6c4e4bb7a18b80
\ No newline at end of file
+https://github.com/ARM-software/CMSIS-DSP/#72d6ccb7bf0ed5c7d3a0fb199d852d59ff31e85b
```
Any reason why this is changed. The previous commit ID was correct. It should be:
https://github.com/ARM-software/CMSIS-DSP/#3a04f817a4380c00ff51f9bcdc6c4e4bb7a18b80

ahaslam2 replied:
Ok. i was getting compiler errors if i updated only no-os. i had to update the other projects too, i thought we should always be on the latest of eatch.   After i updated the sub modules to what is pointed by the current sha, there are no compiler errors, so i i can leave it as ti was.

mphalke commented on .mbedignore:
```diff
@@ -28,3 +28,4 @@ projects/adt7420_console/
 projects/ad559xr_console/
 projects/ad5933_console/
 projects/evb_discovery_firmware/
+projects/ad2s1210_iio/
```
Could we move this above evb_discovery_firmware please? We want discovery firmware to be last in the list purposefully.

mphalke commented on libraries/lvgl.lib:
```diff
@@ -1 +1 @@
-https://github.com/lvgl/lvgl/#7506b61527195fdc96803fe94f5ba141b6e5a9e8
\ No newline at end of file
+https://github.com/lvgl/lvgl/#a65ca267bf9e4a55d6a303aee17cfbe22d464f2b
```
Not sure why this is changed. Previous one was correct. Please use:
https://github.com/lvgl/lvgl/#7506b61527195fdc96803fe94f5ba141b6e5a9e8

mphalke commented on libraries/no-OS.lib:
```diff
@@ -1 +1 @@
-https://github.com/analogdevicesinc/no-OS/#3ade1be92427f544ccf3589eaba3f9e8cb1f2312
+https://github.com/analogdevicesinc/no-OS/#ac149ee97d668c901c8f8e02cdd17b2c8ccb34ea
```
please use: https://github.com/analogdevicesinc/no-OS/#6b024163b1257dbd7c36a72bef693d4194cc1d5a

mphalke commented on libraries/precision-converters-library.lib:
```diff
@@ -1 +1 @@
-https://github.com/analogdevicesinc/precision-converters-library/#f29801b9138cbf56d50f7a619e3b3e0acba02623
\ No newline at end of file
+https://github.com/analogdevicesinc/precision-converters-library/#f1c0aec3510727e4ef3153e217da596c9e39a19a
```
please use: https://github.com/analogdevicesinc/precision-converters-library/#eca26432da174d4110434c91087e07e16ea4d587

mphalke commented on projects/ad2s1210_iio/.gitignore:
```diff
@@ -0,0 +1,12 @@
+*.swp
```
to be consistent with other projects, please use below content in this file:

*.swp
*.profile*
build
.vs/
.visualgdb/
*.sln
*.vgdbproj
*.TMP
*.user
tests/.pytest_cache
tests/__pycache__
tests/func/__pycache__
tests/output
tests/.idea

mphalke commented:
@ahaslam2 Some more questions:

1. What's the plan on adding build and test support for this project? I think we should have them in place. Please refer any existing project and check how are adding build and test support (it's quite straightforward). We need to have ci_build.groovy file in the project directory and then pytest based tests folder.

2. I didn't see anything done w.r.t. eeprom validation and context attribute read/validation in the FW. Have a look at this code here: https://github.com/analogdevicesinc/precision-converters-firmware/blob/main/projects/ad4130_iio/app/ad4130_iio.c#L1692. It is important to have bord info validated and conveyed to client tool so that evaluation software detects the valid board. Have a look at this file created by Phillip which contains the checklist for firmware release. https://github.com/analogdevicesinc/precision-converters-firmware/blob/main/CONTRIBUTING.md. If you want to do that as  a part of new PR, please do so.

3. There are few things which you may not be able to do such as updating ADI product pages, link wiki page to product page, etc. These things are part of release checklist (CONTRIBUTING.md). So I would suggest to prepare a list of items which are not done as per checklist and share with us, so that we know what is pending and if we can do from our side. I would be bit skeptical of the releasing project without this checklist.

4. I assume STM32 work is also planned and that would be added as a part of new PR

I will get back to you if anything is pending once above things are addressed. Thanks.

ahaslam2 commented on projects/ad2s1210_iio/app/ad2s1210_user_config.c:
```diff
@@ -0,0 +1,65 @@
+	},
+	.gpio_res1 = {
+		.number = ARDUINO_UNO_D6,
+		.platform_ops = &mbed_gpio_ops,
```
same comment, it was fixed on the latest patch:
.gpio_res1 = {
		.number = GPIO_RES1,
		.platform_ops = &mbed_gpio_ops

ahaslam2 commented on projects/ad2s1210_iio/makefile:
```diff
@@ -0,0 +1,7 @@
+include ../../clone-repos.mk
```
i tested with mbed-cli and make not with  keil studio compiler

jansunil commented on projects/ad2s1210_iio/README.txt:
```diff
@@ -0,0 +1,65 @@
+DIGITAL-13	->	13  SCLK
+POWER-GND	->	4   GIO
+POWER-3v3	->	116 VIO
+
```
Is the PWM pin D3 included in this list?

jansunil commented on projects/ad2s1210_iio/tests/func/test_ad2s1210_example.py:
```diff
@@ -0,0 +1,70 @@
+iio_device = { 'DEV_AD2S1210': 'ad2s1210',}
+#max angle is 2pi radians
+MAX_EXPECTED_ANGLE = 65535
+MIN_EXPECTED_ANGLE = 0
```
Are we validating the readback value from 0 to full scale? The test result would pass for any value between 0-65535 here.

Generally there's a fixed source of input and the validation happens for a specific value that we already know.
Has there been any discussion on what is the source of input here?

ahaslam2 commented on projects/ad2s1210_iio/app/ad2s1210_user_config.c:
```diff
@@ -0,0 +1,64 @@
+	},
+	.gpio_sample = {
+		.number = GPIO_SAMPLE,
+		.platform_ops = &mbed_gpio_ops,
```
ok, when i saw your comment it showed on a old version of the patch that had GPIO_SAMPLE as the "ARDUINO" pins
so i thought you were referring to that.  ill remove mbed_gpio_ops

ahaslam2 commented on projects/ad2s1210_iio/tests/func/test_ad2s1210_example.py:
```diff
@@ -0,0 +1,70 @@
+    for sample_index in range(0, len(data)):
+        for ch_id in range (0, len(ad2s1210_dev.rx_enabled_channels)):
+            data[sample_index] = chn_data[ad2s1210_dev.rx_enabled_channels[ch_id]]
+            sample_index = sample_index + len(ad2s1210_dev.channel)
```
this comes from the ad777x and ad7134  i would think this was the accepted way,  i can rewrite this

ahaslam2 commented on projects/ad2s1210_iio/tests/func/test_ad2s1210_example.py:
```diff
@@ -0,0 +1,70 @@
+    # Extract the data of enabled channels in rx_enabled_channels
+    for sample_index in range(0, len(data)):
+        for ch_id in range (0, len(ad2s1210_dev.rx_enabled_channels)):
+            data[sample_index] = chn_data[ad2s1210_dev.rx_enabled_channels[ch_id]]
```
i copied from projects/ad777x_iio/tests/func/test_ad777x_example.py

i took it as the "reference code", but ok ill change it.

mphalke commented on projects/ad2s1210_iio/README.txt:
```diff
@@ -0,0 +1,65 @@
+Hardware Setup
+--------------
+Required: SDP-K1, ASZS-BREAKOUT, EVAL-AD2S1210 board and USB cable.
+Plug in the EVAL-AD2S1210 board to the ASZS-BREAKOUT, populate flywires to the
```
looks like typo:
ASZS-BREAKOUT should be ADZS-BRKOUT

ahaslam2 replied:
ill change it

mphalke replied:
looks like same change is required on line 21 as well, typo there as well

ahaslam2 commented:
Change log:

- change ASZS-BREAKOUT to ADZS-BRKOUT in readme

mphalke commented on projects/ad2s1210_iio/app/app_config_mbed.c:
```diff
@@ -0,0 +1,78 @@
+	.spi_clk_pin = SPI_SCK,
+	.spi_miso_pin = SPI_HOST_SDI,
+	.spi_mosi_pin = SPI_HOST_SDO,
+	.use_sw_csb = true
```
any particular reason of using software CSB? As per our experience with Mbed, software CSB adds-up some extra dead time during data transfer, whereas with hardware CSB this dead time is less compared to software CSB. This is expected as we are explicitly toggling CS pin when it is configured as software CSB, while when it is configured as h/w CSB, toggling of CSB happen within Mbed layer, so this extra overhead is less.

ahaslam2 replied:
i need to try that.

mphalke commented on projects/ad2s1210_iio/tests/func/ad2s1210.py:
Is plan to move this file within pyadi-iio repo and use it from there? All driver scripts are placed on pyadi-iio repo.

ahaslam2 replied:
i saw some projects where this file is on the project:
./ad7134_iio/tests/func/ad7134.py
./ad7606_iio/tests/func/ad7606.py
./ad4696_iio/tests/func/ad4696.py
./ad77681_iio/tests/func/ad77681.py

But looking at support in pyadi-iio https://github.com/analogdevicesinc/pyadi-iio/blob/main/adi/ad2s1210.py
i can  get rid of this.

mphalke replied:
yes, we need to move these files into pyadi-iio repo.

ahaslam2 commented:
Change log:

- remove ad2s1210.py use the one in pyadi-iio
- use hw cs instead of sw cs
- fix another ASZS-BREAKOUT -> ASZS-BRKOUT in readme

mphalke commented:
looks good to me unless any comments from @jansunil and @ribdp

jansunil commented on projects/ad2s1210_iio/README.txt:
```diff
@@ -0,0 +1,65 @@
+Hardware Setup
+--------------
+Required: SDP-K1, ADZS-BRKOUT, EVAL-AD2S1210 board and USB cable.
+Plug in the EVAL-AD2S1210 board to the ASZS-BRKOUT, populate flywires to the
```
ASZS-> ADZS

ahaslam2 commented:
changelog:
- fix typo in readme doc
